Organizing calls

sneakzttv opened the pull request:
Removed the many example scripts we have and combined them into 1 centralized script for each protocol ERC20/1155 etc etc.

Users may find this easier to locate and subsequently the methods are a lot cleaner which you can see from the file changes.

This also makes alterations easier as we only need to update our changes in 1 place.

If we like this approach I will continue on and create a script for each protocol.

Update:

After long discussions between myself nikola and rob, we decided on the above approach, i've added a few changes below:

Removed texture scene and merged it into the main scene so we're 1 step closer to a single scene.
![image](https://github.com/ChainSafe/web3.unity/assets/57473220/1c3480c8-ea4a-4f28-9f33-978f0b664a27)

Grouped all of the protocol scripts in the root of the scene under SDK Calls for easy access rather than diving through buttons.
![image](https://github.com/ChainSafe/web3.unity/assets/57473220/c1d3625e-a685-43c3-8cf8-301d2377a4cd)

Updated Unsorted to EVM as it displays nicer.
![image](https://github.com/ChainSafe/web3.unity/assets/57473220/d1a4413c-d306-4dcf-9d16-003ac1922602)

Removed the many behavior scripts, now we just have a centralized script for each protocol, yay.
![image](https://github.com/ChainSafe/web3.unity/assets/57473220/079fbc60-6b44-4890-a2b9-e34363b0c43d)

Updated tests to work with the changes.

Added ecdsa sign, get address, send tx functions and mint and transfer functions for 20/721 & 1155. Created buttons for each function in the main scene.

This merge will make calling the SDK functions simple and easy.
![image](https://github.com/ChainSafe/web3.unity/assets/57473220/2f26dee2-bc28-4b2b-8de3-f6c123ba96fb)


-----------------To Test---------------------
Boot up the login scene to enter into the main scene and check the calls to make sure that everything is functioning as intended.

robGG1997 commented on Packages/io.chainsafe.web3-unity/Runtime/Scripts/EVM/Token/Erc20.cs:
```diff
@@ -1,40 +1,65 @@
+        public static async Task<BigInteger> BalanceOf(string contractAddress, string account)
         {
-            var contract = web3.ContractBuilder.Build(Abi, contractAddress);
+            var contract = Web3Accessor.Web3.ContractBuilder.Build(Abi, contractAddress);
```
If this class isn't going to be part of Samples we should not use `Web3Accessor`

sneakzttv replied:
noted, updated

robGG1997 commented on Packages/io.chainsafe.web3-unity/Runtime/Scripts/EVM/Token/Erc20.cs:
```diff
@@ -1,40 +1,65 @@
+        /// <param name="contractAbi"></param>
+        /// <param name="contractAddress"></param>
+        /// <returns></returns>
+        public static async Task<string> CustomTokenBalance(string contractAbi, string contractAddress)
```
Do these methods need to be static? can't we have an instance and call then from that instance?

sneakzttv replied:
noted, updated

sneakzttv commented:
Removed texture scene and merged it into the main scene so we're 1 step closer to a single scene.
Grouped all of the protocol scripts in the root of the scene under SDK Calls for easy access.
Updated Unsorted to EVM as it displays nicer.
Updated tests to work with the changes.
Removed the many behavior scripts.

-----------------To Test---------------------
Boot up the main scene and check that everything is functioning as intended.

sneakzttv commented:
hm, pipeline tests are failing on 4 instances, fixing.

kantagara commented on Packages/io.chainsafe.web3-unity/Runtime/Scripts/EVM/Token/Evm.cs:
```diff
@@ -0,0 +1,189 @@
+
+namespace Scripts.EVM.Token
+{
+    public class Evm
```
Class itself can then also be static since all of it's methods are static.

sneakzttv replied:
good point, updating.

sneakzttv replied:
done

RyRy79261 commented on Packages/io.chainsafe.web3-unity/Runtime/Scripts/EVM/Token/Erc1155.cs:
```diff
@@ -92,6 +111,93 @@ public static async Task<string> URI(Web3 web3, string contractAddress, string t
+        /// <param name="id"></param>
+        /// <param name="amount"></param>
+        /// <returns></returns>
+        public static async Task<object[]> MintErc1155(Web3 web3, string abi, string contractAddress, int id, int amount)
```
These should not be int parameters, if you are minting an ERC20 style token then the amount could overflow considering the decimals. Additionally, if Chainsafe ERC1155 examples that have the IPFS hash baked into the ID will overflow this parameter.

robGG1997 replied:
so BigInteger, right?

robGG1997 replied:
if so @sneakzttv you can reference OwnerOf fix currently on main

sneakzttv replied:
updated

RyRy79261 commented on Packages/io.chainsafe.web3-unity/Runtime/Scripts/EVM/Token/Erc721.cs:
```diff
@@ -133,5 +148,46 @@ public static async Task<string> URI(Web3 web3, string contractAddress, string t
+        /// <param name="toAccount"></param>
+        /// <param name="tokenId"></param>
+        /// <returns></returns>
+        public static async Task<object[]> TransferErc721(Web3 web3, string contractAddress, string toAccount, int tokenId)
```
Token ID's should be in string format as they aren't used in calculations but only to identify. Additionally, Large token ID's like CS's ERC1155 tokens will overflow

sneakzttv replied:
updated

RyRy79261 commented on Packages/io.chainsafe.web3-unity/Samples~/Web3.Unity/Scripts/Scenes/SampleMain/EVM/EvmCalls.cs:
```diff
@@ -0,0 +1,277 @@
+    #region Contract Call
+
+    private string methodCall = "myTotal";
+    private string abiCall = "[ { \"inputs\": [ { \"internalType\": \"uint8\", \"name\": \"_myArg\", \"type\": \"uint8\" } ], \"name\": \"addTotal\", \"outputs\": [], \"stateMutability\": \"nonpayable\", \"type\": \"function\" }, { \"inputs\": [], \"name\": \"myTotal\", \"outputs\": [ { \"internalType\": \"uint256\", \"name\": \"\", \"type\": \"uint256\" } ], \"stateMutability\": \"view\", \"type ...
```
This abi is the same as the one on line 33

RyRy79261 replied:
@kantagara @robGG1997 So just wanna check if this is alright, the code here, @sneakzttv says is a template for using our examples so would be an exception to the rule when avoiding duplication of variables. Does that make sense?

sneakzttv replied:
same as below

robGG1997 replied:
We need to have a single point of reference as much as possible unless our use case explicitly demands we use different ones

RyRy79261 commented on Packages/io.chainsafe.web3-unity/Samples~/Web3.Unity/Scripts/Scenes/SampleMain/EVM/EvmCalls.cs:
```diff
@@ -0,0 +1,277 @@
+    private string contractAddressArray = "0x5244d0453A727EDa96299384370359f4A2B5b20a";
+    private string abiArray = "[{\"inputs\":[{\"internalType\":\"address[]\",\"name\":\"_addresses\",\"type\":\"address[]\"}],\"name\":\"setStore\",\"outputs\":[],\"stateMutability\":\"nonpayable\",\"type\":\"function\"},{\"inputs\":[{\"internalType\":\"uint256\",\"name\":\"\",\"type\":\"uint256\"}],\"name\":\"bought\",\"outputs\":[{\"internalType\":\"address\",\"name\":\"\",\"type\":\"address\"} ...
+    private string methodArrayGet = "getStore";
+    private string methodArraySend = "getStore";
```
The contents here are the same as line 51

sneakzttv replied:
fixed

RyRy79261 commented on Packages/io.chainsafe.web3-unity/Samples~/Web3.Unity/Scripts/Scenes/SampleMain/EVM/EvmCalls.cs:
```diff
@@ -0,0 +1,277 @@
+        
+    private string methodSend = "addTotal";
+    private string abiSend = "[ { \"inputs\": [ { \"internalType\": \"uint8\", \"name\": \"_myArg\", \"type\": \"uint8\" } ], \"name\": \"addTotal\", \"outputs\": [], \"stateMutability\": \"nonpayable\", \"type\": \"function\" }, { \"inputs\": [], \"name\": \"myTotal\", \"outputs\": [ { \"internalType\": \"uint256\", \"name\": \"\", \"type\": \"uint256\" } ], \"stateMutability\": \"view\", \"type ...
+    private string contractAddressSend = "0x7286Cf0F6E80014ea75Dbc25F545A3be90F4904F";
```
This would be network dependent

sneakzttv replied:
goerli, yes

RyRy79261 commented on src/UnitySampleProject/Assets/Samples/web3.unity SDK/2.5.0/Web3.Unity Samples/Scripts/Scenes/SampleMain/EVM/EvmCalls.cs:
```diff
@@ -0,0 +1,277 @@
+    # region Contract Send
+        
+    private string methodSend = "addTotal";
+    private string abiSend = "[ { \"inputs\": [ { \"internalType\": \"uint8\", \"name\": \"_myArg\", \"type\": \"uint8\" } ], \"name\": \"addTotal\", \"outputs\": [], \"stateMutability\": \"nonpayable\", \"type\": \"function\" }, { \"inputs\": [], \"name\": \"myTotal\", \"outputs\": [ { \"internalType\": \"uint256\", \"name\": \"\", \"type\": \"uint256\" } ], \"stateMutability\": \"view\", \"type ...
```
This ABI appears in other files

sneakzttv replied:
yeah, it's just there to show users what variables to add directly

robGG1997 replied:
I see your point @sneakzttv, I guess it's okay for samples but we should have a single point of ref for ABIs imo

sneakzttv replied:
updated

juans-chainsafe reviewed (changes requested):
Ok! I finish the testing, here is the status:

- WebGL build with Web3Auth ✅  except https://github.com/ChainSafe/web3.unity/issues/714
- Web3Auth in Unity editor ✅ 
- Wallet connect in Unity editor ✅ 

❌ Failing:

- All721Nft (we should remove this, the endpoint is no longer supported in game-api)
- All1155Nft (we should remove this, the endpoint is no longer supported in game-api)
- Multicall
```
InvalidOperationException: No service for type 'ChainSafe.Gaming.MultiCall.IMultiCall' has been registered.
Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService (System.IServiceProvider provider, System.Type serviceType) (at <1ba52dc0b103422eba91ad87a1aa9fe3>:0)
Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T] (System.IServiceProvider provider) (at <1ba52dc0b103422eba91ad87a1aa9fe3>:0)
ChainSafe.Gaming.MultiCall.MultiCallExtensions.MultiCall (ChainSafe.Gaming.Web3.Web3 web3) (at <c5a8911fc36e44c88a494f120980ea1f>:0)
EvmCalls.MultiCall () (at Assets/Samples/web3.unity SDK/2.5.0/Web3.Unity Samples/Scripts/Scenes/SampleMain/EVM/EvmCalls.cs:257)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <130809ae6f984869a6663c878f16e3f3>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

sneakzttv commented:
fixed

juans-chainsafe reviewed (approved):
Multicall working now! ✅ 
All721Nft and All1155Nft are failing ❌ but they will be fixed in a different PR

RyRy79261 reviewed (approved):
Code looks good, Sneakz ran test to confirm the BigInteger works in the calls
